application/apps/indexer/addons/bufread/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+pub struct BufReader<R> {
+    reader: R,
+    buffer: DeqBuffer,
+    min: usize,
+}
+
+impl<R> BufReader<R> {
+    /// Creates a new reader with the given maximum space and minimum bytes to be buffered ahead.
+    pub fn new(max: usize, min: usize, reader: R) -> Self {
+        BufReader {
+            reader,
+            buffer: DeqBuffer::new(max),
+            min,
+        }
+    }
+
+    /// Returns the total capacity of the inner buffer.
+    pub fn capacity(&self) -> usize {
+        self.buffer.capacity()
+    }
+
+    /// Returns the number of currently available bytes of the inner buffer.
+    pub fn len(&self) -> usize {
+        self.buffer.read_available()
+    }
+
+    /// Answers if the inner buffer is currently empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the current slice to read from of the inner buffer.
+    pub fn buffer(&self) -> &[u8] {
+        self.buffer.read_slice()
+    }
+}
+
+impl<R: Read> Read for BufReader<R> {
+    fn read(&mut self, buffer: &mut [u8]) -> Result<usize> {
+        if self.buffer.write_available() < self.min {
+            self.buffer.flush();
+        }
+
+        let size = self.fill_buf()?.read(buffer)?;
+        self.consume(size);
+        Ok(size)
+    }
+}
+
+impl<R: Read> BufRead for BufReader<R> {
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        if self.buffer.read_available() < self.min {
+            self.buffer.flush();
+        }
+
+        if let Ok(size) = self.reader.read(self.buffer.write_slice()) {
+            self.buffer.write_done(size);
+        }
+
+        Ok(self.buffer.read_slice())
+    }
+
+    fn consume(&mut self, size: usize) {
+        self.buffer.read_done(size);
+    }
+}
+
+impl<R: Seek> Seek for BufReader<R> {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        let result: u64;
+
+        if let SeekFrom::Current(pos) = pos {
+            result = self
+                .reader
+                .seek(SeekFrom::Current(pos - self.len() as i64))?;
+        } else {
+            result = self.reader.seek(pos)?;
+        }
+
+        self.buffer.clear();
+        Ok(result)
+    }
+}
+
+/// A fixed buffer that could be used as a byte dequeue.
+pub struct DeqBuffer {
+    slice: Box<[u8]>,
+    start: usize,
+    end: usize,
+}
+
+impl DeqBuffer {
+    /// Creates a new buffer with the given maximum space.
+    ///
+    /// # Safety
+    ///
+    /// The buffer needs to be written to before read from.
+    pub fn new(size: usize) -> Self {
+        let mut vec = Vec::with_capacity(size);
+        let cap = vec.capacity();
+
+        // SAFETY: Force the length of the vector to its capacity
+        // instead of using 'resize' for performance reasons.
+        unsafe {
+            vec.set_len(cap);
+        }
+
+        DeqBuffer {
+            slice: vec.into(),
+            start: 0,
+            end: 0,
+        }
+    }
+
```

Comment:
couldn't it make sense for us not to forcefully shorten the vector in case more space was reserved than we need?
of course it would alter the semantics of the function, but it's internal API

Reply:
Actually it's more a forcefully inflate of the vector, because after Vec::with_capacity "the returned vector has the minimum capacity specified, the vector will have a zero length". Instead of 'Vec::set_len' same can be achieved by 'Vec::resize', but with a significant performance drawback.